### PR TITLE
Standalone sweep once softmax kernel w/ ckProfiler

### PR DESCRIPTION
**Pull request by rosenrodt**
- Add profile_normalization() for standalone softmax profiling. Can extend to standalone layernorm/batchnorm as well
- Add int8 softmax instances. However, it's not for real use cases as quantized data doesn't scale linearly after applying exp() so is neither able to requantize nor dequantize in any trivial way
- Add "sweep once" kernel. Will try to load data from global mem just once when problem size permits. 130% speedup (30% gain) is observed.

Sweep-once softmax kernel supporting limited reduction size; basically limited by block tile size
```
$ build/bin/ckProfiler softmax 1 0 1 0 1 --length 512 512 512 --reduce 2
Perf: 0.546594 ms, 982.211 GB/s, DeviceReduceSoftmax<256,M_C4_S1,K_C64_S8,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>
```
Generic softmax supporting ~unlimited~ large reduction size
```
Perf: 0.705778 ms, 760.679 GB/s, DeviceReduceSoftmax<256,M_C4_S1,K_C64_S8,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>
```

Reduce add bandwidth:
```
$ build/bin/ckProfiler reduce --half -D 1,512,512,512 -R 3 -O 0 -C 1 -v 0 1 10
Best Perf: 0.281361 ms, 955.925 GB/s
```

### CODE REVIEW

File: include/ck/tensor_operation/gpu/device/device_softmax.hpp [hunk 1]
```diff
@@ -170,21 +204,24 @@ struct DeviceSoftmax : public BaseOperator
...
-                                                      void* out_dev)
+                                                      void* out_dev) override
     {
         return std::make_unique<Argument>(inLengths,
                                           inStrides,
                                           reduceDims,
-                                          alpha,
-                                          beta,
+                                          *static_cast<AccDataType*>(alpha),
+                                          *static_cast<AccDataType*>(beta),
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/device/device_softmax.hpp:216]: We probably need to document the specification used here.  Reduction will align with this. 

And we might need more discussion, since:

1. The cuDNN and MIOpen specification for `scaling parameter` is more complicated than just using `AccDataType as type of alpha, beta`
2. For some Reduction operation, like MAX/MIN/AMAX `AccDataType` is usually the same as `InOutDataType` for best accessing performance.  In that case, whether it is most reasonable to use `AccDataType` for alpha/beta when `AccDataType` is half_t or int8_t
3. And another problem is, whether we really need to use `alpha` and `beta` for scaling and bias when Reduction is `MAX/MIN/AMAX` ?

Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/device_softmax.hpp:216]: > The cuDNN and MIOpen specification for scaling parameter is more complicated than just using AccDataType as type of alpha, beta

What would it be? I couldn't find a lot concrete info on [MIOpen doc](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/softmax.html#). And cuDNN [documents about scaling precision type](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#scaling-parameters) but not the accumulation type, either.

> For some Reduction operation, like MAX/MIN/AMAX AccDataType is usually the same as InOutDataType for best accessing performance. In that case, whether it is most reasonable to use AccDataType for alpha/beta when AccDataType is half_t or int8_t
> And another problem is, whether we really need to use alpha and beta for scaling and bias when Reduction is MAX/MIN/AMAX 

True that there's no point promoting to higher precision when it's max/min reduction. I can make max reduction part of softmax normalization leverage it. The last part of your question about reduction alpha/beta scaling is confusing to me. Is the question for reduction itself or softmax normalization?

Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/device_softmax.hpp:216]: > > The cuDNN and MIOpen specification for scaling parameter is more complicated than just using AccDataType as type of alpha, beta
> 
> What would it be? I couldn't find a lot concrete info on [MIOpen doc]

(https://rocmsoftwareplatform.github.io/MIOpen/doc/html/softmax.html#). And cuDNN [documents about scaling precision type](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#scaling-parameters) but not the accumulation type, either.
> 
> > For some Reduction operation, like MAX/MIN/AMAX AccDataType is usually the same as InOutDataType for best accessing performance. In that case, whether it is most reasonable to use AccDataType for alpha/beta when AccDataType is half_t or int8_t
> > And another problem is, whether we really need to use alpha and beta for scaling and bias when Reduction is MAX/MIN/AMAX
> 
> True that there's no point promoting to higher precision when it's max/min reduction. I can make max reduction part of softmax normalization leverage it. The last part of your question about reduction alpha/beta scaling is confusing to me. Is the question for reduction itself or softmax normalization?

I think before  3) is clarified,  use `AccDataType *` is most reasonable.  Since: 
>  When in case of  Reduction MAX/MIN/AMAX,  we can assume alpha and beta will not be actually used, so it does not matter whether to use `AccDataType` or higher precision type for alpha/beta
>  In case of Reduction `Add/Mul/NORM1/NORM2`,  `AccDataType` is already required to be of higher precision, so it is not a problem to use `AccDataType` for alpha, beta

Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/device_softmax.hpp:216]: So the question is mostly about reduction itself, not normalization? What actions do I need to take here?

File: profiler/include/profile_normalization_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                                std::vector<index_t> in_length,
+                                std::vector<index_t> in_strides,
+                                std::vector<index_t> reduce_dims,
+                                AccDataType alpha,
+                                AccDataType beta,
+                                NormType norm_type)
+{
+    Tensor<InDataType> in = in_strides.empty() ? Tensor<InDataType>(in_length)
+                                               : Tensor<InDataType>(in_length, in_strides);
+    Tensor<OutDataType> out(in);
```
[the remark this answers is not in the document]
Comment (rosenrodt) [reply profiler/include/profile_normalization_impl.hpp]: Actually, it should not assume input and output are of the same precision. Fixed in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/295/commits/bbddd01719e958c0c293d648966fd0e74dec59db

**Review by qianfengz — APPROVED** [review]

**Review by qianfengz — APPROVED** [review]
